scarlet2/initialization.py

```diff
@@ -0,0 +1,331 @@
+        for obs, model in zip((observation,), models):
+            for i in range(perimeter_pixels.shape[1]):
+                spectrum = perimeter_pixels[:, i]
+                psf_model = obs.frame.psf()
+                # psf_model = jnp.expand_dims(psf_model, axis=0)
+                psf_peak = psf_model.max(axis=(1, 2))
+                spectrum /= psf_peak
+                spectrum_avg += spectrum
+            spectrum_avg /= perimeter_pixels.shape[1]
+            spectra.append(spectrum_avg)
+        spectra_box = spectra[0]
+
+        # now check the SNR and spectrum correlations
+        noise_avg = jnp.mean(perimeter_noise_rms, axis=1) * noise_thresh
+        snr = spectra_box / noise_avg
+        spec_corr = jnp.dot(spectra_box, peak_spec) / jnp.dot(
+            jnp.sqrt(jnp.dot(spectra_box, spectra_box)),
+            jnp.sqrt(jnp.dot(peak_spec, peak_spec)),
+        )
+
+        # check if SNR is below threshold or spectrum correlation is below threshold
+        if snr.all() < 1 or spec_corr < corr_thresh:
+            converged = True
+        # check if flux on boundaries increased with larger size
+        elif (spectra_box >= peak_spec).any() or (spectra_box >= spectra_prev).any():
+            converged = True
+        # increase box/morpgology size
+        else:
+            spectra_prev = spectra_box
+            sigma += 1
+        # for debugging
+        if sigma > 5:
+            break
+
+    # normalise the morphology between 0 and 1
+    morph = morph()
+    morph = (morph - jnp.min(morph)) / (jnp.max(morph) - jnp.min(morph))
+    return morph
+
+
+def init_morphology(
+    obs,
+    center,
+    psf_sigma=1,
+    noise_thresh=100,
+    corr_thresh=0.8,
+    max_size=32,
+    components=1,
+):
+    """Initialize the morphology of the sources.
+    Parameters
+    ----------
+    obs : `~scarlet.observation.Observation` instance
+    center : tuple (int, int) central pixel of the sources
+
+    Returns
+    -------
+    morph : np.ndarray morphology of the source normalised to (0,1)
+    """
+    morph = init_simple_morph(obs, center, psf_sigma, noise_thresh, corr_thresh)
+    if morph.shape[0] <= max_size:
+        return morph
+    # fit for a more complex morphology for bigger sources
+    else:
+        bx = morph.shape[0]
+        cutout = cut_square_box(obs.data[0], center, bx)
+        sx, sy, th = fit_morph_params(cutout, center, bx)
+        morph = create_gaussian_array(bx, sx, sy, th)
+        morph = np.array(morph, dtype=np.float32)  # convert to numpy to make mutable
+        rows, cols = morph.shape
+        central_row = rows // 2
+        central_col = cols // 2
+        morph[central_row, central_col] = 1.1  # extra brightening central pixel
+        morph = (morph - np.min(morph)) / (np.max(morph) - np.min(morph))
+        if bx > 30 and components == 2:
+            morph2 = create_gaussian_array(
+                bx, 1.25, 1.25, 0
+            )  # create a second component as a gaussian blob
```

Comment:
I don't understand the heuristics that are used here. Why is this factor 1.1 there, what justifies the criterion for having 2 components (it looks like it's a size argument, which is indirectly a SNR argument), and what sets the size of the secondary component?

Reply:
Removed fitting routine